xarray/core/computation.py

```diff
@@ -926,6 +926,86 @@ def earth_mover_distance(first_samples,
         return apply_array_ufunc(func, *args, dask=dask)
 
 
+def dot(*args, **kwargs):
+    """ dot(*arrays, dims=None)
+
+    einsum for xarray object.
+
+    Parameters
+    ----------
+    arrays: arrays to compute
+    dims: tuple of strings, optional
+        Along which dimensions to be summed over.
+        If None is provided, then all the common dimensions are summed over.
+
+    Returns
+    -------
+    dot: same type to input.
+
+    Examples
+    --------
+
+    >>> da_a = xr.DataArray(np.arange(3 * 4).reshape(3, 4), dims=['a', 'b'])
+    >>> da_b = xr.DataArray(np.arange(3 * 4 * 5).reshape(3, 4, 5),
+                            dims=['a', 'b', 'c'])
+    >>> da_c = xr.DataArray(np.arange(5 * 6).reshape(5, 6), dims=['c', 'd'])
+
+    >>> dot(da_a, da_b, dims=['a', 'b']).dims
+    ('c', )
+    >>> dot(da_a, da_b, dims=['a']).dims
+    ('b', 'c')
+    >>> dot(da_a, da_b, da_c, dims=['b', 'c']).dims
+    ('a', 'd')
+    """
+    dims = kwargs.pop('dims', None)
+    arrays = args
+    if dims is None and isinstance(args[-1], (list, tuple, basestring)):
+        dims = args[-1]
+        arrays = args[:-1]
```

Comment:
I think it is better to require specifying `dims` with a keyword argument.

Reply:
Our previous `dot` does not require `dim`. This assumes to sum over along all the common dimensions.
I think `dim=None` is not surprising.

Reply:
I agree, the default `dims=None` should be OK. I meant that `dims` should be a keyword only argument, not a required argument.

Here you are supporting `xr.dot(a, b, 'x')`, where `'x'` denotes a dimension. I would require writing `xr.dot(a, b, dim='x')` or omitting `dim` altogether.